### ISSUE TITLE
No logging of compiled code

### ISSUE THREAD

**Issue by MaxOstrowski** [pytorch/pytorch#119186]
### 🐛 Describe the bug

After setting the loglevel like so:
`torch._logging.set_logs(all=logging.WARNING, output_code=True)`
I do get the compiled code output for some function I compile but not for others.

The code below features 2 functions that can be compiled.
Compiling function `test` results in a detailed logging of the compilation process and the resulting code.
Compiling function `test2` results in no log at all.

### Error logs

No error output is provided and the function is executed as normal. Just the output_code is missing.

### Minified repro

```
import numpy as np
import torch
import logging


def test(foo, bar):
    return np.count_nonzero(foo) + np.count_nonzero(bar)

def test2(foo, bar):
    for clause in foo:
        for i in range(0, bar.shape[1]):
            res = np.logical_xor(clause, bar[i])
            if np.count_nonzero(res) == 1:
                bar[i] |= res
    return bar


foo = np.random.choice([True]*5 + [False]*5, size=(10,10))
bar = np.random.choice([True] + [False]*10, size=(10,10))

torch._logging.set_logs(all=logging.WARNING, output_code=True)

compiled = torch.compile(test2)
compiled(foo, bar)import numpy as np
import torch
import logging


def test(foo, bar):
    return np.count_nonzero(foo) + np.count_nonzero(bar)

def test2(foo, bar):
    for clause in foo:
        for i in range(0, bar.shape[1]):
            res = np.logical_xor(clause, bar[i])
            if np.count_nonzero(res) == 1:
                bar[i] |= res
    return bar


foo = np.random.choice([True]*5 + [False]*5, size=(10,10))
bar = np.random.choice([True] + [False]*10, size=(10,10))

torch._logging.set_logs(all=logging.WARNING, output_code=True)

compiled = torch.compile(test2) # works for test
compiled(foo, bar)
```

### Versions

PyTorch version: 2.2.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: version 3.16.3
Libc version: glibc-2.31

Python version: 3.11.5 (main, Sep 11 2023, 13:23:44) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-92-generic-x86_64-with-glibc2.31
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Byte Order:                         Little Endian
Address sizes:                      39 bits physical, 48 bits virtual
CPU(s):                             8
On-line CPU(s) list:                0-7
Thread(s) per core:                 2
Core(s) per socket:                 4
Socket(s):                          1
NUMA node(s):                       1
Vendor ID:                          GenuineIntel
CPU family:                         6
Model:                              140
Model name:                         11th Gen Intel(R) Core(TM) i7-1185G7 @ 3.00GHz
Stepping:                           1
CPU MHz:                            1800.000
CPU max MHz:                        4800,0000
CPU min MHz:                        400,0000
BogoMIPS:                           3609.60
Virtualization:                     VT-x
L1d cache:                          192 KiB
L1i cache:                          128 KiB
L2 cache:                           5 MiB
L3 cache:                           12 MiB
NUMA node0 CPU(s):                  0-7
Vulnerability Gather data sampling: Mitigation; Microcode
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l2 invpcid_single cdp_l2 ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves split_lock_detect dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req avx512vbmi umip pku ospke avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid movdiri movdir64b fsrm avx512_vp2intersect md_clear flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.2.0
[conda] _tflow_select             2.3.0                       mkl  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py311h5eee18b_1  
[conda] mkl_fft                   1.3.8           py311h5eee18b_0  
[conda] mkl_random                1.2.4           py311hdb19cb5_0  
[conda] numpy                     1.23.5          py311h08b1b3b_1  
[conda] numpy-base                1.23.5          py311hf175353_1  
[conda] pytorch                   2.2.0              py3.11_cpu_0    pytorch
[conda] pytorch-mutex             1.0                         cpu    pytorch
[conda] tensorflow                2.12.0          mkl_py311h34a0fa1_0  
[conda] tensorflow-base           2.12.0          mkl_py311he5f8e37_0

**Comment by ezyang** [comment]
That's because we didn't compile it. If you TORCH_LOGS=dynamo you will find out there was a graph break inside of a for loop. And the graph break arises because of the `np.count_nonzero(res) == 1` conditional.

**Comment by MaxOstrowski** [comment]
Thanks a lot for your help, I appreciate it, but I can not reproduce what you said.
The documentation:
https://pytorch.org/docs/stable/logging.html
says that `torch._logging.set_logs(all=logging.WARNING`
should include and superseed `TORCH_LOGS=dynamo`.
But even with explicitly setting `TORCH_LOGS=dynamo` I do not get any logging output. No hint that no code was compiled and no hint on the graph break.

The documentation on `torch.compile`
https://pytorch.org/docs/stable/generated/torch.compile.html
does not say anything about graph breaks and states that all functions are compiled.

In the official tutorial:
https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html
it is mentioned that:

> Primarily, the advantage of torch.compile lies in its ability to handle arbitrary Python code with minimal changes to existing code.
> One case that torch.compile can handle that other compiler solutions struggle with is data-dependent control flow (the if x.sum() < 0: line below).

and then the following example function with control flow is shown:
```
def f1(x, y):
    if x.sum() < 0:
        return -y
    return y
```

1. Why do I not see any logging for my function ?
2. Is there a difference to the control flow shown in the example and if yes how can I find out what to change.

**Comment by ezyang** [comment]
> The documentation:
[pytorch.org/docs/stable/logging.html](https://pytorch.org/docs/stable/logging.html)
says that torch._logging.set_logs(all=logging.WARNING

As far as I can tell, it does not. No where on the page is there a code snippet including logging.WARNING. What specific lines are you referring to?

> says that torch._logging.set_logs(all=logging.WARNING should include and superseed TORCH_LOGS=dynamo.

It doesn't say anything about this, but if you try setting both you will get a warning saying that the envvar takes precedence. (Feel free to send a PR clarifying this.)

> The documentation on torch.compile
[pytorch.org/docs/stable/generated/torch.compile.html](https://pytorch.org/docs/stable/generated/torch.compile.html)
does not say anything about graph breaks and states that all functions are compiled.

But... it doesn't say that?

> Concretely, for every frame executed within the compiled region, we will **attempt** to compile it and cache the compiled result on the code object for future use.

> Primarily, the advantage of torch.compile lies in its ability to handle arbitrary Python code with minimal changes to existing code.

Yes, because when we see Python code that is too complicated, we break and fall back to running regular Python. When this happens, you will not get compilation related logs, as we are not compiling.

This being said, I am always looking for ways to improve our docs. Please feel free to send PRs that clarify wording if you have ideas for how to make things more clear!

**Comment by MaxOstrowski** [comment]
Thank you very much for this elaborate explanation.
I must have misread the documentation, you are right that it does not explicitly say so
(also setting `TORCH_LOGS=dynamo` didn't give me any warning that envvar takes over, but maybe my system is just misconfigured).

I can see now that my function wasn't compiled as it was deemed too complicated.
As this is my first time using pytorch I would kindly ask if it would be possible to give a short explanation why a certain function couldn't be compiled in the logs ?

The function from the tutorial:
```
def f1(x, y):
    if x.sum() < 0:
        return -y
    return y
```
works very nice, and I do get translated code, and my function:
```
def test2(foo, bar):
    for clause in foo:
        for i in range(0, bar.shape[1]):
            res = np.logical_xor(clause, bar[i])
            if np.count_nonzero(res) == 1:
                bar[i] |= res
    return bar
```
is silently seen as too complex. It would be very nice, especially for non-experts with this, to show an explanation which part of the function is deemed to complex and maybe why. To me they look the same, and it could be

- the for loops
- the access to `shape`
- the if (also the example function has a very similar if)
- changing of the input parameter `bar`

Thank you very much

**Comment by ezyang** [comment]
The problem is the data dependent access in count_nonzero. We are not a magic compiler, we need a straight line trace through the code. In your code, we can't tell if you go in the branch or not. Your second program to not use if is closer but you probably need to enable something like `capture_scalar_outputs` or something

**Comment by MaxOstrowski** [comment]
Thank you again for patience. Unfortunately I wasn't able to find any information about `capture_scalar_outputs` in the pytorch documentation. I found some mentions in github issues, but nothing concrete. Could you give me a hint/reference how to use it and what it does?

I thought that `compile` is able to partially compile python code into parallizable GPU code when possible leaving the control flow at python level. So I expected np.logical_xor to be translated and executed on the CPU while conditional execution is steared on the python level. I think I was wrong about this and have to do more research on how to execute my algorithm on a CUDA device.

I still think that some logging feedback "why" the translation process is aborted would be beneficial for the user.
Thanks a lot.

**Comment by ezyang** [comment]
torch._dynamo.config.capture_scalar-outputs = True

> I thought that compile is able to partially compile python code into parallizable GPU code when possible leaving the control flow at python level

yes and this is done by graph break
